### PR TITLE
Make it possible to 'crash' the ephemeral file system

### PR DESCRIPTION
**Pull request by mneedham**

### CODE REVIEW

**Comment by lutovich** [comment]
@mneedham Is this different from `EphemeralFileSystemAbstraction#snapshot()` ?

**Comment by mneedham** [comment]
@lutovich I think it's slightly different as in the `crash` case we want to change the existing FileSystem whereas with snapshot we create a new one. I'm going to add some tests against it to show what we want to do then hopefully it'll be clearer.

**Comment by mneedham** [comment]
Rebased to run again now that the debian fix is in there

**Comment by chrisvest** [comment]
@mneedham Please test add thorough tests for it. Bugs in EphemeralFSA are annoying to hunt down, and can make otherwise fine test and product code appear flaky and buggy in the builds.

**Comment by mneedham** [comment]
@chrisvest do you think the tests I've added are sufficient or what other ones should I add?

File: community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstractionTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        EphemeralFileSystemAbstraction fs = new EphemeralFileSystemAbstraction();
+
+        StoreChannel channel = fs.open( aFile, "rw" );
+        writeLong( channel, 1111 );
+
+        // when
+        fs.shutdown();
+
+        // then
+        assertEquals(0, fs.open( aFile, "r" ).size());
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstractionTest.java]: This test seems to contradict its name

File: community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstraction.java [hunk 2]
```diff
@@ -900,6 +912,18 @@ void open( EphemeralFileChannel channel )
...
+        void force()
+        {
+            forcedBuffer = fileAsBuffer.copy();
+            forcedSize = size;
+        }
+
+        void crash()
+        {
+            fileAsBuffer = forcedBuffer.copy();
+            size = forcedSize;
```
Comment (chrisvest) [inline community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstraction.java]: The write method synchronises on the `fileAsBuffer` field and updates the size, so I think the locking in here might need an overhaul.

**Comment by chrisvest** [comment]
Tests that races reads, writes, forces and crashes would be good. If force races with a write, it'd be good to know that the internal invariants are maintained in the DynamicByteBuffer copy.

File: community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstractionTest.java (hunk 1 above)
Comment (mneedham) [inline community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstractionTest.java]: d'oh, my mistake. Will fix.

**Comment by mneedham** [comment]
@digitalstain can you have a look at this one? We now have tests for single threaded usage.

**Comment by MishaDemianenko** [comment]
@mneedham header missing

File: community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstractionTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        ByteBuffer readBuffer = allocateDirect( 8 );
+        readChannel.read( readBuffer );
+        readBuffer.flip();
+        return readBuffer;
+    }
+
+    private void writeLong( StoreChannel channel, long value ) throws IOException
+    {
+        ByteBuffer buffer;
+        buffer = allocateDirect( 8 );
```
Comment (MishaDemianenko) [inline community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstractionTest.java]: assignment should be merged with declaration

**Pull request merged**